File: generics/DogsAndRobots.java
// generics/DogsAndRobots.java
// (c)2021 MindView LLC: see Copyright.txt
// We make no guarantees that this code is fit for any purpose.
// Visit http://OnJava8.com for more book information.
// Java中并无（直接的）潜在类型机制
import pets.Dog;

// 表演的狗
class PerformingDog extends Dog implements Performs {
    @Override
    public void speak() {
        System.out.println("Woof!");
    }

    @Override
    public void sit() {
        System.out.println("Sitting");
    }

    public void reproduce() {
    }
}

class Robot implements Performs {
    @Override
    public void speak() {
        System.out.println("Click!");
    }

    @Override
    public void sit() {
        System.out.println("Clank!");
    }

    public void oilChange() {
    }
}

// Communicate 沟通、传达
class Communicate { // 去泛型 版本，见下个示例。
    // perform()方法不需要使用泛型来工作，它可以被简单地指定为接受一个Performs 对象
    public static <T extends Performs> void perform(T performer) {
        performer.speak();
        performer.sit();
    }
}

/**
 *  潜在类型机制
 *
 *  java不具备 潜在类型机制
 *  但可以通过接口 实现 直接潜在类型
 *
 *  这种方式不够泛化，我们只能调用 接口中定义的方法，而不可能是“任意方法”，
 *  而且，后期维护新增功能，只能新增继承的接口 或 在已继承的接口中新增方法定义。
 *
 *
 *
 *  这里，泛型不是必须的，因此，可改写成以下版本：
 *  generics/SimpleDogsAndRobots.java
 *
 * @see SimpleDogsAndRobots
 */
public class DogsAndRobots {
    public static void main(String[] args) {
        // 使用沟通类 给 狗 传达 表演指令
        Communicate.perform(new PerformingDog());
        // 使用沟通类 给 机器人 传达 表演指令
        Communicate.perform(new Robot());
    }
}
/* Output:
Woof!
Sitting
Click!
Clank!
*/

File: staticchecking/dogsandrobots.go
// staticchecking/dogsandrobots.go
// (c)2021 MindView LLC: see Copyright.txt
// We make no guarantees that this code is fit for any purpose.
// Visit http://OnJava8.com for more book information.
package main

import "fmt"

// Dog
// Go没有Class关键字，但你可以用上面的形式等价地创建基类：
//		把你平时要创建的类，改为创建struct，
//		struct中则存放着数据字段（本例中没有数据字段）。
type Dog struct{}
// 每个方法都由func关键字开头，然后，再把方法附着到类上————————用圆括号将对象引用包起来，对象引用可以是任何标识符，但这里用了this，以提醒你这就像C++或Java中的this一样。
// 然后，定义函数中剩下的部分，就和你在Go中定义任何其他函数一样。
func (this Dog) speak()     { fmt.Printf("Arf!\n") }
func (this Dog) sit()       { fmt.Printf("Sitting\n") }
func (this Dog) reproduce() {}

type Robot struct{}

func (this Robot) speak()     { fmt.Printf("Click!\n") }
func (this Robot) sit()       { fmt.Printf("Clank!\n") }
func (this Robot) oilChange() {}

// Go中也没有继承，所以这种“面向对象”的形式相当原始，可能这也是我不愿在这门语言上花费更多时间的主要原因。不过，组合的方式很直观。

/*
perform()函数使用了潜在类型机制：参数的具体类型并不重要，只要它含有speak()方法和sit()方法即可。
这里，在行内匿名地创建了interface，正如你在perform()函数的参数列表中所见。
*/
func perform(speaker interface {
	speak()
	sit()
}) {
	speaker.speak()
	speaker.sit()
}

/*
从main()中可以看出，perform()确实并不关心其参数的具体类型，只要可以在该参数上调用speak()和sit()即可。
不过，就像C++的模板函数一样，会在编译时进行参数类型校验。
语法Dog{}和Robot{}，创建了匿名的Dog和Robot的struct。
*/
func main() {
	perform(Dog{})
	perform(Robot{})
}

/* 输出：
   Arf!
   Sitting
   Click!
   Clank!
*/
